File: .tmp_sc4013/java_22/3956_maximum_sum_of_m_non_overlapping_subarrays_i.java
// CONFIG class=Solution method=maxSum types=None
// LeetCode 3956 - Maximum Sum of M Non-Overlapping Subarrays I
// https://leetcode.com/problems/maximum-sum-of-m-non-overlapping-subarrays-i/

import java.util.ArrayList;
import java.util.List;

class Solution {
    public long maxSum(int[] nums, int m, int l, int r) {
        int n = nums.length;
        long[] prefix = new long[n + 1];
        for (int i = 0; i < n; i++) prefix[i + 1] = prefix[i] + nums[i];
        long[] dp = new long[n + 1];
        long bestSelected = -(1L << 62);
        for (int count = 1; count <= m; count++) {
            long[] next = dp.clone();
            var deque = new ArrayList<Integer>();
            for (int end = 1; end <= n; end++) {
                int addIndex = end - l;
                if (addIndex >= 0) {
                    long value = dp[addIndex] - prefix[addIndex];
                    while (deque.size() > 0) {
                        int last = deque.get(deque.size() - 1);
                        if (dp[last] - prefix[last] > value) break;
                        deque.remove(deque.size() - 1);
                    }
                    deque.add(addIndex);
                }
                int minIndex = end - r;
                while (deque.size() > 0 && deque.get(0) < minIndex) deque.remove(0);
                if (deque.size() > 0) {
                    long candidate = prefix[end] + dp[deque.get(0)] - prefix[deque.get(0)];
                    if (candidate > next[end]) next[end] = candidate;
                    if (candidate > bestSelected) bestSelected = candidate;
                }
                if (next[end - 1] > next[end]) next[end] = next[end - 1];
            }
            dp = next;
        }
        return bestSelected;
    }
}

File: .tmp_rs4013/rust_check22/3956_maximum_sum_of_m_non_overlapping_subarrays_i.rs
struct Solution;
// LeetCode 3956 - Maximum Sum of M Non-Overlapping Subarrays I
// https://leetcode.com/problems/maximum-sum-of-m-non-overlapping-subarrays-i/

use std::collections::VecDeque;

impl Solution {
    pub fn max_sum(nums: Vec<i32>, m: i32, l: i32, r: i32) -> i64 {
        let n = nums.len();
        let mut prefix = vec![0i64; n + 1];
        for i in 0..n {
            prefix[i + 1] = prefix[i] + nums[i] as i64;
        }
        let mut dp = vec![0i64; n + 1];
        let mut best_selected = -(1i64 << 62);
        for _count in 1..=m {
            let mut next = dp.clone();
            let mut deque: VecDeque<usize> = VecDeque::new();
            for end in 1..=n {
                let add_index = end as i32 - l;
                if add_index >= 0 {
                    let add_index = add_index as usize;
                    let value = dp[add_index] - prefix[add_index];
                    while let Some(&last) = deque.back() {
                        if dp[last] - prefix[last] > value {
                            break;
                        }
                        deque.pop_back();
                    }
                    deque.push_back(add_index);
                }
                let min_index = end as i32 - r;
                while let Some(&front) = deque.front() {
                    if (front as i32) < min_index {
                        deque.pop_front();
                    } else {
                        break;
                    }
                }
                if let Some(&front) = deque.front() {
                    let candidate = prefix[end] + dp[front] - prefix[front];
                    if candidate > next[end] {
                        next[end] = candidate;
                    }
                    if candidate > best_selected {
                        best_selected = candidate;
                    }
                }
                if next[end - 1] > next[end] {
                    next[end] = next[end - 1];
                }
            }
            dp = next;
        }
        best_selected
    }
}

fn main() {}
